### PR TITLE
Update pydantic and ophyd_async to latest versions

### LINKED ISSUE #454: Update pydantic and ophyd_async
Following https://github.com/DiamondLightSource/dodal/issues/679 we should make sure we're on the latest version of `pydantic` and `ophyd_async`

## Acceptance Criteria
- `mx-bluesky` uses the latest version of `pydantic` and `ophyd_async`

### PR DESCRIPTION
**Pull request by DominicOram**
Fixes #454

Most changes should just be like for like behaviour. The only functional change should is that we now by default allow extra fields when constructing models. However, when doing the conversion in the external facing API we error if there are extra fields. Previously we would disallow extra fields but optionally change it at runtime in `from_json`. In pydantic 2 it is harder to change at runtime and we probably want to allow the flexibility by default so we can easily convert between parameter models. Ideally this would mean we can get rid of `from_json` but this proved a bit tricky so I spun into https://github.com/DiamondLightSource/mx-bluesky/issues/457.

Also needs https://github.com/DiamondLightSource/dodal/pull/779 from dodal

To test:
* Confirm unit tests pass

### CODE REVIEW

**Review by olliesilvester — CHANGES_REQUESTED** [review]
Looks OK I think. The linting is failing because of the recent change to `detector.py` in `dodal`. We just need to remove the line 

```
beam_xy_converter=DetectorDistanceToBeamXYConverter(
                self.det_dist_to_beam_converter_path
            ),
```

from `class RotationExperiment`

**Comment by DominicOram** [comment]
> Looks OK I think. The linting is failing because of the recent change to detector.py in dodal. We just need to remove the line...

You mean like https://github.com/DiamondLightSource/mx-bluesky/pull/456/commits/04a006e38b309061983d9eb9fa23e55fafdcf631 ;)

**Review by olliesilvester — APPROVED** [review]
OK, looks good!

**Comment by olliesilvester** [comment]
Yeah, I wasn't 100% on why a few of the changes were made but they looked sensible enough

File: src/mx_bluesky/hyperion/parameters/components.py [hunk 1]
```diff
@@ -31,16 +37,6 @@ def _parse(cls, version):
             return version
```
[the remark this answers is not in the document]
Comment (DominicOram) [reply src/mx_bluesky/hyperion/parameters/components.py:37]: Thanks, spun into https://github.com/DiamondLightSource/mx-bluesky/issues/461

**Pull request merged**